### PR TITLE
geo/geomfn: implement ST_Snap

### LINKED ISSUE #49040: geo/geomfn: implement ST_Snap({geometry,geometry,float8})
Implement `ST_Snap` on arguments {geometry,geometry,float8}, which should adopt [PostGIS behaviour](https://postgis.net/docs/ST_Snap.html).

_Observers: Please react to this issue if you need this functionality._

For Geometry builtins, please do the following:
* Ideally add a relevant helper function in [`pkg/geo/geomfn`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo/geomfn) (parse and output related functions can go in [`pkg/geo`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo)). Add exhaustive unit tests here - you can run through example test cases and make sure that PostGIS and CRDB return the same result within a degree of accuracy (1cm for geography).
  * When using GEOS, you can reference the [C API](https://github.com/libgeos/geos/blob/master/capi/geos_c.h.in) for which functions are available. Unfortunately, Windows is not currently supported when using GEOS.
* Create a new builtin that references this function in [`pkg/sql/sem/builtins/geo_builtins.go`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sem/builtins/geo_builtins.go). Note that we currently do not support optional arguments, so we define functions that have optional arguments once without the optional argument (using the default value in the optional position), and once with the optional argument.
* Modify the tests in [`pkg/sql/logictest/testdata/logic_test/geospatial`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/geospatial) to call this functionality at least once. You can call `make testbaselogic FILES='geospatial' TESTFLAGS='-rewrite'` to regenerate the output. Tests here should just ensure the builtin is linked end to end (your exhaustive unit tests go the above mentioned packages!).
* Ensure the documentation is regenerated by calling `make buildshort`. You can also play with it by calling `./cockroach demo --empty` afterwards.
* Submit your PR - make sure to follow the guidelines from [creating your first PR](https://wiki.crdb.io/wiki/spaces/CRDB/pages/181633464/Your+first+CockroachDB+PR]).

You can follow #48552 for an example PR.

**The following additional guidance has been issued on implementing this function:**

Implemented by GEOS

<sub>:robot: This issue was synced with a spreadsheet by [gsheets-to-github-issues](https://github.com/cockroachlabs/gsheet-to-github-issues) by [otan](https://github.com/otan) on 2021-02-21T20:42:26Z. Changes to titles, body and labels may be overwritten.</sub>

### PR DESCRIPTION
**Pull request by busiest-beaver**
This patch implements ST_Snap.

Resolves cockroachdb/cockroach#49040

Release justification: low risk, high benefit changes to existing functionality
Release note (sql change): ST_Snap is now available.

### CODE REVIEW

File: pkg/geo/geos/geos.h [hunk 1]
```diff
@@ -192,6 +192,9 @@ CR_GEOS_Status CR_GEOS_RelatePattern(CR_GEOS* lib, CR_GEOS_Slice a, CR_GEOS_Slic
 CR_GEOS_Status CR_GEOS_VoronoiDiagram(CR_GEOS* lib, CR_GEOS_Slice g, CR_GEOS_Slice env,
                                       double tolerance, int onlyEdges, CR_GEOS_String* ret);
 
+// Snap
```
Comment (otan) [inline pkg/geo/geos/geos.h:195]: nit: remove this comment

File: pkg/sql/logictest/testdata/logic_test/geospatial [hunk 2]
```diff
@@ -5768,3 +5768,18 @@ SRID=4326;POINT (359 90)
 SRID=4326;POINT (-179 -23)
 SRID=4326;LINESTRING (2 2, 5 -60, -160 0)
 SRID=4326;POLYGON ((354 -10, 6 -10, 0 20, 354 -10), (3 2, 359 -1, 1 -5, 3 2))
+
```
Comment (otan) [inline pkg/sql/logictest/testdata/logic_test/geospatial:5771]: nit: extra new line

File: pkg/sql/logictest/testdata/logic_test/geospatial [hunk 3]
```diff
@@ -5768,3 +5768,18 @@ SRID=4326;POINT (359 90)
...
+SELECT ST_AsText(
+    ST_Snap(poly,line, ST_Distance(poly, line)*1.25)
+  ) AS polysnapped
+FROM (SELECT
+  ST_GeomFromText('MULTIPOLYGON(
+    (( 26 125, 26 200, 126 200, 126 125, 26 125 ),
+      ( 51 150, 101 150, 76 175, 51 150 )),
+      (( 151 100, 151 200, 176 175, 151 100 )))') As poly,
+       ST_GeomFromText('LINESTRING (5 107, 54 84, 101 100)') As line
+	) As _;
```
Comment (otan) [inline pkg/sql/logictest/testdata/logic_test/geospatial:5783]: i tihnk this `AS _` should bre replaced by `tbl (poly, line)` to make it easier to parse.
or use `SELECT ST_asText(poly, ...) FROM [SELECT ST_GeomFromText(...)]`

File: pkg/geo/geomfn/snap.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
+// Copyright 2020 The Cockroach Authors.
```
Comment (otan) [inline pkg/geo/geomfn/snap.go:1]: nit: change dates to 2021 here and below

**Review by otan — COMMENTED** [review]
thanks! some small things

File: pkg/sql/logictest/testdata/logic_test/geospatial (hunk 3 above)
Comment (alsohas) [reply pkg/sql/logictest/testdata/logic_test/geospatial:5783]: would this work?
```
query T
SELECT ST_AsText(
    ST_Snap(poly,line, ST_Distance(poly, line)*1.25)
) AS polysnapped
FROM (SELECT
    ST_GeomFromText('MULTIPOLYGON(
      (( 26 125, 26 200, 126 200, 126 125, 26 125 ),
      ( 51 150, 101 150, 76 175, 51 150 )),
      (( 151 100, 151 200, 176 175, 151 100 )))') As poly,
    ST_GeomFromText('LINESTRING (5 107, 54 84, 101 100)') As line
  ) tbl(poly, line);
```

Comment (otan) [reply pkg/sql/logictest/testdata/logic_test/geospatial:5783]: If it passes ya

**Comment by otan** [comment]
looks like a missing `make bazel-generate`

**Review by otan — APPROVED** [review]
bors r+

**Pull request merged**